File: python/pyjse/ast/nodes.py
"""Concrete AST node implementations for JSE.

Following the design in docs/regular.md:
- SymbolNode: Represents a symbol reference like $x
- ArrayNode: Represents [operator, args...] function call
- ObjectNode: Represents {"$operator": value} operator call
- QuoteNode: Represents quoted (unevaluated) expression
- LambdaNode: Represents lambda function with closure
"""

from typing import Any, TYPE_CHECKING
from pyjse.ast.base import AstNode
from pyjse.types import JseValue

if TYPE_CHECKING:
    from pyjse.env import Env


class SymbolNode(AstNode):
    """Represents a symbol reference like $x.

    When applied, looks up the symbol in the call-time environment.
    """

    def __init__(self, name: str, env: 'Env') -> None:
        """Initialize symbol node.

        Args:
            name: Symbol name (e.g., "$x")
            env: Construct-time environment
        """
        super().__init__(env)
        self._name = name

    def apply(self, env: 'Env') -> JseValue:
        """Look up symbol in call-time environment.

        Args:
            env: Call-time environment for lookup

        Returns:
            The bound value

        Raises:
            NameError: If symbol is not found
        """
        value = env.resolve(self._name)
        if value is None:
            raise NameError(f"Symbol '{self._name}' not found")
        return value


class ArrayNode(AstNode):
    """Represents an array expression.

    Can be either:
    - A function call: [operator, args...]
    - A regular array: [elements...]
    """

    def __init__(self, elements: list[JseValue], env: 'Env') -> None:
        """Initialize array node.

        Args:
            elements: List of elements (may include AST nodes)
            env: Construct-time environment
        """
        super().__init__(env)
        self._elements = elements

    def apply(self, env: 'Env') -> JseValue:
        """Evaluate array node.

        If first element is a SymbolNode, treats as function call.
        Otherwise evaluates all elements.

        Args:
            env: Call-time environment

        Returns:
            Function call result or evaluated array
        """
        if not self._elements:
            return []

        first = self._elements[0]

        # Check if this is a function call form
        if isinstance(first, SymbolNode):
            # Look up the functor
            functor = env.resolve(first._name)
            if functor is None:
                raise NameError(f"Unknown operator: {first._name}")

            # Special forms: don't evaluate all arguments
            # $def, $defn, $lambda need unevaluated symbols/expressions
            special_forms = ('$def', '$defn', '$lambda', '$quote')
            if first._name in special_forms:
                # Pass arguments unevaluated (functor will handle)
                if callable(functor):
                    return functor(env, *self._elements[1:])
                return functor

            # Regular functors: evaluate arguments first
            evaluated_args = [env.eval(arg) for arg in self._elements[1:]]

            # Call the functor
            if callable(functor):
                return functor(env, *evaluated_args)
            return functor

        # Regular array - evaluate all elements
        return [env.eval(elem) for elem in self._elements]


class ObjectNode(AstNode):
    """Represents an object with key-value pairs.

    Form: {key: value, ...}
    """
    def __init__(self, dict: dict[str, JseValue], env: 'Env') -> None:
        """Initialize object node.
        """
        super().__init__(env)
        self._dict = dict
        
    def apply(self, env: 'Env') -> JseValue:
        """Evaluate object node.
        """
        return {k: env.eval(v) for k, v in self._dict.items()}


class ObjectExpressionNode(AstNode):
    """Represents an object expression with operator key.

    Form: {"$operator": value, "meta": ...}
    """

    def __init__(
        self,
        operator: str,
        value: JseValue,
        metadata: dict[str, JseValue],
        env: 'Env'
    ) -> None:
        """Initialize object node.

        Args:
            operator: The operator symbol (e.g., "$add")
            value: The operand value
            metadata: Additional metadata keys
            env: Construct-time environment
        """
        super().__init__(env)
        self._operator = operator
        self._value = value
        self._metadata = metadata

    def apply(self, env: 'Env') -> JseValue:
        """Evaluate object node as operator call.

        Args:
            env: Call-time environment

        Returns:
            Operator result
        """
        # Look up the functor
        functor = env.resolve(self._operator)
        if functor is None:
            raise NameError(f"Unknown operator: {self._operator}")

        # Evaluate the value
        evaluated_value = env.eval(self._value)

        # For special operators like $expr, value might need wrapping
        if self._operator == "$expr":
            args = [evaluated_value]
        elif isinstance(evaluated_value, list):
            args = evaluated_value
        else:
            args = [evaluated_value]

        # Call the functor
        if callable(functor):
            return functor(env, *args)
        return functor


class QuoteNode(AstNode):
    """Represents a quoted (unevaluated) expression.

    Form: ["$quote", expr] or {"$quote": expr}
    Returns the expression without evaluation.
    """

    def __init__(self, value: JseValue, env: 'Env') -> None:
        """Initialize quote node.

        Args:
            value: The quoted expression (returned as-is)
            env: Construct-time environment
        """
        super().__init__(env)
        self._value = value

    def apply(self, env: 'Env') -> JseValue:
        """Return quoted value without evaluation.

        Args:
            env: Call-time environment (ignored)

        Returns:
            The quoted value as-is
        """
        return self._value


class LambdaNode(AstNode):
    """Represents a lambda function with closure.

    Captures construct-time environment for static scoping.
    When applied, creates new environment with closure as parent.
    """

    def __init__(
        self,
        params: list[str],
        body: JseValue,
        closure_env: 'Env'
    ) -> None:
        """Initialize lambda node.

        Args:
            params: Parameter names (e.g., ["$x", "$y"])
            body: Function body expression
            closure_env: Environment to capture (construct-time)
        """
        super().__init__(closure_env)
        self._params = params
        self._body = body
        self._closure_env = closure_env

    def apply(self, env: 'Env', *args: JseValue) -> JseValue:
        """Apply lambda with arguments.

        Creates new environment with closure as parent (static scoping).

        Args:
            env: Call-time environment
            *args: Argument values to bind to parameters

        Returns:
            Result of evaluating body

        Raises:
            ValueError: If argument count doesn't match parameter count
        """
        if len(args) != len(self._params):
            raise ValueError(
                f"Lambda expects {len(self._params)} args, got {len(args)}"
            )

        # Create new environment for this call
        # Parent is closure_env (static scoping!)
        from pyjse.env import Env
        call_env = Env(parent=self._closure_env)

        # Bind parameters to arguments
        for param, arg in zip(self._params, args):
            call_env.set(param, arg)

        # Evaluate body in call environment
        return call_env.eval(self._body)


class LiteralNode(AstNode):
    """Represents a literal value (number, string, bool, null).

    Returns itself when applied (no evaluation needed).
    """

    def __init__(self, value: JseValue, env: 'Env') -> None:
        """Initialize literal node.

        Args:
            value: The literal value
            env: Construct-time environment
        """
        super().__init__(env)
        self._value = value

    def apply(self, env: 'Env') -> JseValue:
        """Return literal value.

        Args:
            env: Call-time environment (ignored)

        Returns:
            The literal value
        """
        return self._value

File: gojse/ast/nodes.go
package ast

import (
	"fmt"
	"strings"

	"github.com/MarchLiu/jse/gojse/functors"
)

// MetaSetter is an interface for setting/clearing metadata context
// This avoids circular import between ast and jse packages
type MetaSetter interface {
	SetMeta(meta map[string]interface{})
	ClearMeta()
}

// LiteralNode represents a literal value (numbers, strings, bool, null).
type LiteralNode struct {
	value interface{}
	env   interface{}
}

func NewLiteralNode(value interface{}, env interface{}) *LiteralNode {
	return &LiteralNode{value: value, env: env}
}

func (n *LiteralNode) Apply(env interface{}) (interface{}, error) {
	return n.value, nil
}

func (n *LiteralNode) ToJSON() interface{} {
	return n.value
}

func (n *LiteralNode) GetEnv() interface{} {
	return n.env
}

// SymbolNode represents a variable reference.
type SymbolNode struct {
	name string
	env  interface{}
}

func NewSymbolNode(name string, env interface{}) *SymbolNode {
	return &SymbolNode{name: name, env: env}
}

func (n *SymbolNode) Apply(env interface{}) (interface{}, error) {
	// TODO: Implement symbol resolution through Env interface
	return nil, fmt.Errorf("symbol resolution not yet implemented: %s", n.name)
}

func (n *SymbolNode) ToJSON() interface{} {
	return n.name
}

func (n *SymbolNode) GetEnv() interface{} {
	return n.env
}

func (n *SymbolNode) Name() string {
	return n.name
}

// ArrayNode represents a function call or regular array.
type ArrayNode struct {
	elements []AstNode
	env      interface{}
}

func NewArrayNode(elements []AstNode, env interface{}) *ArrayNode {
	return &ArrayNode{elements: elements, env: env}
}

func (n *ArrayNode) Apply(env interface{}) (interface{}, error) {
	if len(n.elements) == 0 {
		return []interface{}{}, nil
	}

	// Check if first element is a symbol (function call)
	first := n.elements[0]
	if sym, ok := first.(*SymbolNode); ok && isSymbol(sym.Name()) {
		return n.applyFunctionCall(env, sym.Name())
	}

	// Regular array - evaluate all elements
	result := make([]interface{}, len(n.elements))
	for i, el := range n.elements {
		v, err := el.Apply(env)
		if err != nil {
			return nil, err
		}
		result[i] = v
	}
	return result, nil
}

func (n *ArrayNode) ToJSON() interface{} {
	result := make([]interface{}, len(n.elements))
	for i, el := range n.elements {
		result[i] = el.ToJSON()
	}
	return result
}

func (n *ArrayNode) GetEnv() interface{} {
	return n.env
}

func (n *ArrayNode) applyFunctionCall(env interface{}, symbol string) (interface{}, error) {
	// Special forms that don't evaluate arguments
	specialForms := map[string]bool{
		"$quote": true,
	}

	// Get rest of arguments (elements after the first)
	rest := n.rest()

	if specialForms[symbol] {
		// Pass unevaluated arguments
		args := make([]interface{}, len(rest))
		for i, el := range rest {
			args[i] = el.ToJSON()
		}
		return n.applyFunctor(env, symbol, args)
	}

	// Regular functors - evaluate arguments first
	evaluated := make([]interface{}, len(rest))
	for i, el := range rest {
		v, err := el.Apply(env)
		if err != nil {
			return nil, err
		}
		evaluated[i] = v
	}
	return n.applyFunctor(env, symbol, evaluated)
}

func (n *ArrayNode) rest() []AstNode {
	if len(n.elements) <= 1 {
		return []AstNode{}
	}
	return n.elements[1:]
}

func (n *ArrayNode) applyFunctor(env interface{}, symbol string, args []interface{}) (interface{}, error) {
	envImpl, ok := env.(functors.Env)
	if !ok {
		return nil, fmt.Errorf("env does not implement functor.Env interface")
	}
	return envImpl.ApplyFunctor(symbol, args)
}

// ObjectNode represents a key-value pair object.
type ObjectNode struct {
	dict map[string]AstNode
	env  interface{}
}

func NewObjectNode(dict map[string]AstNode, env interface{}) *ObjectNode {
	return &ObjectNode{dict: dict, env: env}
}

func (n *ObjectNode) Apply(env interface{}) (interface{}, error) {
	result := make(map[string]interface{})
	for key, node := range n.dict {
		v, err := node.Apply(env)
		if err != nil {
			return nil, err
		}
		result[key] = v
	}
	return result, nil
}

func (n *ObjectNode) ToJSON() interface{} {
	result := make(map[string]interface{})
	for key, node := range n.dict {
		result[key] = node.ToJSON()
	}
	return result
}

func (n *ObjectNode) GetEnv() interface{} {
	return n.env
}

// ObjectExpressionNode represents {"$operator": value, ...}.
type ObjectExpressionNode struct {
	operator string
	value    AstNode
	metadata map[string]interface{} // Metadata associated with this expression
	env      interface{}
}

func NewObjectExpressionNode(operator string, value AstNode, metadata map[string]interface{}, env interface{}) *ObjectExpressionNode {
	return &ObjectExpressionNode{operator: operator, value: value, metadata: metadata, env: env}
}

func (n *ObjectExpressionNode) Apply(env interface{}) (interface{}, error) {
	// Special handling for $expr
	if n.operator == "$expr" {
		return n.value.Apply(env)
	}

	// Set metadata context before calling functor
	if metaSetter, ok := env.(MetaSetter); ok {
		metaSetter.SetMeta(n.metadata)
	}

	var result interface{}
	var resultErr error
	if n.operator == "$pattern" || n.operator == "$query" {
		// Special handling for $pattern and $query - pass unevaluated JSON
		jsonValue := n.value.ToJSON()
		result, resultErr = n.applyFunctor(env, n.operator, []interface{}{jsonValue})
	} else {
		// For other operators, evaluate and apply
		evaluated, err := n.value.Apply(env)
		if err != nil {
			// Clear metadata context on error
			if metaSetter, ok := env.(MetaSetter); ok {
				metaSetter.ClearMeta()
			}
			return nil, err
		}
		result, resultErr = n.applyFunctor(env, n.operator, []interface{}{evaluated})
	}

	// Clear metadata context after functor call
	if metaSetter, ok := env.(MetaSetter); ok {
		metaSetter.ClearMeta()
	}

	return result, resultErr
}

func (n *ObjectExpressionNode) applyFunctor(env interface{}, symbol string, args []interface{}) (interface{}, error) {
	envImpl, ok := env.(functors.Env)
	if !ok {
		return nil, fmt.Errorf("env does not implement functor.Env interface")
	}
	return envImpl.ApplyFunctor(symbol, args)
}

func (n *ObjectExpressionNode) ToJSON() interface{} {
	result := map[string]interface{}{n.operator: n.value.ToJSON()}
	// Include metadata in JSON representation
	for k, v := range n.metadata {
		result[k] = v
	}
	return result
}

func (n *ObjectExpressionNode) GetEnv() interface{} {
	return n.env
}

// QuoteNode returns unevaluated expression.
type QuoteNode struct {
	value AstNode
	env   interface{}
}

func NewQuoteNode(value AstNode, env interface{}) *QuoteNode {
	return &QuoteNode{value: value, env: env}
}

func (n *QuoteNode) Apply(env interface{}) (interface{}, error) {
	return n.value.ToJSON(), nil
}

func (n *QuoteNode) ToJSON() interface{} {
	return n.value.ToJSON()
}

func (n *QuoteNode) GetEnv() interface{} {
	return n.env
}

// LambdaNode represents a closure with static scoping.
type LambdaNode struct {
	params     []string
	body       AstNode
	closureEnv interface{}
}

func NewLambdaNode(params []string, body AstNode, closureEnv interface{}) *LambdaNode {
	return &LambdaNode{
		params:     params,
		body:       body,
		closureEnv: closureEnv,
	}
}

func (n *LambdaNode) Apply(env interface{}) (interface{}, error) {
	// Lambda evaluation returns a lambda object
	result := map[string]interface{}{
		"__lambda__": true,
		"params":     n.params,
		"body":       n.body.ToJSON(),
	}
	return result, nil
}

func (n *LambdaNode) ToJSON() interface{} {
	return "<lambda>"
}

func (n *LambdaNode) GetEnv() interface{} {
	return n.closureEnv
}

// Helper functions

func isSymbol(s string) bool {
	return strings.HasPrefix(s, "$") && !strings.HasPrefix(s, "$$")
}

func unescapeSymbol(s string) string {
	if strings.HasPrefix(s, "$$") {
		return s[1:]
	}
	return s
}
